LIP-25

avsetsin opened the pull request.

infloop committed:
feat: add IStakingModule changes

infloop committed:
feat: add other changes

AlexanderLukin committed:
feat: first version of LIP for SR 2.0

AlexanderLukin committed:
refactor: minor cleanups

Remove not implemented ToDos.

AlexanderLukin committed:
feat: links to HackMD docs

avsetsin commented on LIPS/lip-25.md:
```diff
@@ -0,0 +1,1987 @@
+
+In the current implementation of the Curated modules, key vetting (a process of making keys depositable) occurs through the DAO, and in particular EasyTrack motion, which the operator initiates after submitting keys. DSM changes aim to improve the process of key vetting to be able to work without the governance approval and to accommodate future permissionless modules.
+
+The current VEBO mechanism only processes validator exits in response to user withdrawal requests. This limitation hinders the protocol's ability to manage validator exits proactively, especially for permissionless modules. From the Staking Router's side, it is proposed to consider the module's share when prioritizing validators for exit.
```
```suggestion
The current VEBO mechanism only requests validators to exit in response to user withdrawal requests. This limitation hinders the protocol's ability to manage validator exits proactively, especially for permissionless modules. From the Staking Router's side, it is proposed to consider the module's share when prioritizing validators for exit.
```
I'm not sure what the term "process" means here, but without EIP-7002 it can only request validators to exit

AlexanderLukin replied:
Good point, I agree! Corrected this phrase.

AlexanderLukin committed:
refactor: update sections with boosted exit requests

avsetsin commented on LIPS/lip-25.md:
```diff
@@ -0,0 +1,1987 @@
+
+### 1. `IStakingModule` interface changes
+
+To support the new [automated keys vetting process](https://hackmd.io/@lido/rJrTnEc2a#Automated-Vetting) in the Deposit Security Module and [Boosted Exit Requests](https://hackmd.io/@lido/rJrTnEc2a#Automated-Vetting) in the Validator Exit Bus Oracle, the `IStakingModule` interface should endure a number of changes.
```
```suggestion
To support the new [automated keys vetting process](https://hackmd.io/@lido/rJrTnEc2a#Automated-Vetting) in the Deposit Security Module and [Boosted Exit Requests](https://hackmd.io/@lido/BJXRTxMRp#Boosted-Exit-Requests1) in the Validator Exit Bus Oracle, the `IStakingModule` interface should endure a number of changes.
```
Wrong link

AlexanderLukin replied:
I've corrected this link and other incorrect links. Thanks!

avsetsin commented on LIPS/lip-25.md:
```diff
@@ -0,0 +1,1987 @@
+
+To support the new [automated keys vetting process](https://hackmd.io/@lido/rJrTnEc2a#Automated-Vetting) in the Deposit Security Module and [Boosted Exit Requests](https://hackmd.io/@lido/rJrTnEc2a#Automated-Vetting) in the Validator Exit Bus Oracle, the `IStakingModule` interface should endure a number of changes.
+
+All the code in this interface assumes the Solidity v0.8.9 syntax.
```
Not sure it's worth focusing on, it's just an interface. Besides NOR is written in 0.4.24 solidity and respects the IStakingModule interface.

avsetsin commented on LIPS/lip-25.md:
```diff
@@ -0,0 +1,1987 @@
+
+### 1. `IStakingModule` interface changes
+
+To support the new [automated keys vetting process](https://hackmd.io/@lido/rJrTnEc2a#Automated-Vetting) in the Deposit Security Module and [Boosted Exit Requests](https://hackmd.io/@lido/rJrTnEc2a#Automated-Vetting) in the Validator Exit Bus Oracle, the `IStakingModule` interface should endure a number of changes.
```
I would focus on unvetting (https://hackmd.io/@lido/rJrTnEc2a#Unvetting)  instead of automatic vetting, because we don't do automatic vetting in this iteration. Only maintain existing vetting via governance and optimistic vetting

AlexanderLukin replied:
Really important note, thanks! Corrected as you suggested.

avsetsin commented on LIPS/lip-25.md:
```diff
@@ -0,0 +1,1987 @@
+///     Deposit Contract. This value is a cumulative counter: even when the validator goes into
+///     EXITED state this counter is not decreasing
+/// @return depositableValidatorsCount number of validators in the set available for deposit
+function getNodeOperatorSummary(uint256 _nodeOperatorId) external view returns (
```
Consider adding a note that the method is backwards compatible

AlexanderLukin replied:
I've added the new special "Backward compatibility notes" subsection to all spec sections where it made sense.

avsetsin commented on LIPS/lip-25.md:
```diff
@@ -0,0 +1,1987 @@
+function updateExitedValidatorsCount(
+    bytes calldata _nodeOperatorIds,
+    bytes calldata _exitedValidatorsCounts
+) external;
```
We don't make changes here. There's just a typo being corrected

AlexanderLukin replied:
Agree. Removed this method from the spec.

AlexanderLukin committed:
refactor: merge 3 subsections of 4-th section into one

- Merge 4.4, 4.5 and 4.6 sections into new "Deposits frequency limits" section and add more details about implementation.
- Add several notes about new key vetting logic.

avsetsin commented on LIPS/lip-25.md:
```diff
@@ -0,0 +1,1983 @@
+
+#### 2.1. Module's share consideration for validator exits
+
+Staking Router now should consider module's share when prioritizing validators for exit. Support of this improvement in the `StakingRouter` contract requires several changes:
```
It's more correct to say that VEBO is responsible for prioritization, but target share and exit threshold params are stored in SR

AlexanderLukin replied:
Yes, agree. I've updated this section with a much more detailed explanation. Your note has also been taken into account.

avsetsin commented on LIPS/lip-25.md:
```diff
@@ -0,0 +1,1983 @@
+    uint64 minDepositBlockDistance;
+}
+
+struct StakingModuleCache {
+    address stakingModuleAddress;
+    uint24 stakingModuleId;
+    uint16 stakingModuleFee;
+    uint16 treasuryFee;
+    uint16 stakeShareLimit;
+    StakingModuleStatus status;
+    uint256 activeValidatorsCount;
+    uint256 availableValidatorsCount;
+}
```
Consider do not pay attention to changes in internal structures

AlexanderLukin replied:
OK, I've removed internal structures from code examples.

avsetsin commented on LIPS/lip-25.md:
```diff
@@ -0,0 +1,1983 @@
+* @param _maxDepositsPerBlock the maximum number of validators that can be deposited in a single block
+* @param _minDepositBlockDistance the minimum distance between deposits in blocks
+*/
+function addStakingModule(
```
Consider leaving only the names (or interfaces) of the external (!) methods that are changed and more explanation in words how it should work. Now quite a lot of code is in LIP and only a part of it changes. Code inserts could probably be useful somewhere to expose individual thoughts, but bringing absolutely every change to the LIP seems unnecessary

AlexanderLukin replied:
Yes, in general, I agree with this suggestion. The text of the LIP has been substantially rewritten. Many text explanations that cover the logic of the implemented code have been added. Many internal methods and structures with not very important updates have been removed from code examples.

avsetsin commented on LIPS/lip-25.md:
```diff
@@ -0,0 +1,1983 @@
+
+#### 2.2. Contract version upgrade
+
+For correct migration to the new version of the `StakingRouter` contract, the existing `initialize` external method should be updated, and the new `finalizeUpgrade_v2` external method should be implemented. The existing `ZeroAddress` error type should be transformed to the new `ZeroAddressLido` error type. Also, the new `ZeroAddressAdmin` error type should be added.
```
Consider adding a link to the LIP that describes the convention for initialisation and upgrade finalisation methods
https://github.com/lidofinance/lido-improvement-proposals/blob/feat/lip-24/LIPS/lip-10.md

AlexanderLukin replied:
Good idea, thanks! I've added this link to all appropriate sections.

avsetsin commented on LIPS/lip-25.md:
```diff
@@ -0,0 +1,1983 @@
+
+#### 2.2. Contract version upgrade
+
+For correct migration to the new version of the `StakingRouter` contract, the existing `initialize` external method should be updated, and the new `finalizeUpgrade_v2` external method should be implemented. The existing `ZeroAddress` error type should be transformed to the new `ZeroAddressLido` error type. Also, the new `ZeroAddressAdmin` error type should be added.
```
> The existing `ZeroAddress` error type should be transformed to the new `ZeroAddressLido` error type. Also, the new `ZeroAddressAdmin` error type should be added.

It's bytecode optimisations, not sure it should be mentioned in the LIP

AlexanderLukin replied:
OK, removed mentions of these errors from the 2.2 section.

avsetsin commented on LIPS/lip-25.md:
```diff
@@ -0,0 +1,1983 @@
+
+#### 2.3. Keys vetting improvements
+
+Support of new key vetting logic requires implementation of the new `decreaseStakingModuleVettedKeysCountByNodeOperator` external method. Also, the new `STAKING_MODULE_UNVETTING_ROLE` constant should be defined.
```
> Also, the new `STAKING_MODULE_UNVETTING_ROLE` constant should be defined.

Basically what we're proposing here is to add `STAKING_MODULE_UNVETTING_ROLE` role and assign it to a DSM

AlexanderLukin replied:
OK, added this note to the 2.3 section.

avsetsin commented on LIPS/lip-25.md:
```diff
@@ -0,0 +1,1983 @@
+
+#### 2.4. New target limit modes
+
+The `StackingRouter` contract should support more than two target limit modes. To implement this the following changes are needed:
```
Stacking -> Staking (in several places in the document)

AlexanderLukin replied:
Fixed, thanks.

avsetsin commented on LIPS/lip-25.md:
````diff
@@ -0,0 +1,1983 @@
+}
+```
+
+#### 2.5. New methods for getting deposit blocks
````
```suggestion
#### 2.5. New methods for getting module deposit limits
```

avsetsin replied:
It is important that we migrate these params from DSM to SR, since each module may have different ones

AlexanderLukin replied:
OK, changed the title and added the relevant notes to this section.

avsetsin commented on LIPS/lip-25.md:
```diff
@@ -0,0 +1,1983 @@
+
+#### 2.7. Excluded methods for module pausing
+
+Methods, events and constants related to Staking Module pausing logic should be removed from the `StakingRouter` contract:
```
And it's proposed to have pause for all deposits on DSM side and unvetting. The motivation is that in the current design if one permisionless operator frontruns deposits than DSM pauses the whole module. We need to change this reaction to a more specific one, from a module to a specific operator.

AlexanderLukin replied:
OK, I've added more details about the purpose of the new pause to the 2.6 section. Also, I've added the link to the new DSM spec, where much more details about this logic are provided.

avsetsin commented on LIPS/lip-25.md:
```diff
@@ -0,0 +1,1983 @@
+
+#### 3.3. New reward distribution engine
+
+The new external permissionless `distributeReward` method should be implemented. This method should distribute all accumulated module rewards among node operators based on the latest accounting report.
+
+Rewards can be distributed after node operators' statistics are updated until the next reward is transferred to the module during the next oracle frame.
```
More input would be helpful here. The main motivation is to leave in the 3d phase of AO report only the update of operator data in modules to have more or less the same size of transactions

AlexanderLukin replied:
Yes, agree with you. I've added much more detail about the new logic of the 3rd phase of the Accounting Oracle report to this section.

avsetsin commented on LIPS/lip-25.md:
```diff
@@ -0,0 +1,1983 @@
+    _updateRewardDistributionState(RewardDistributionState.TransferredToModule);
+}
+
+function onExitedAndStuckValidatorsCountsUpdated() external {
+    _auth(STAKING_ROUTER_ROLE);
+    _updateRewardDistributionState(RewardDistributionState.ReadyForDistribution);
+}
+
+function onExitedAndStuckValidatorsCountsUpdated() external {
+    _auth(STAKING_ROUTER_ROLE);
+    _updateRewardDistributionState(RewardDistributionState.ReadyForDistribution);
+}
```
duplicate

AlexanderLukin replied:
Fixed, thanks!

AlexanderLukin committed:
feat: add more text clarifications

Describe the proposed changes with more text.
Add links to HackMD off-chain specs.

AlexanderLukin committed:
fix: links

- Correct all links;
- Minor cleanups in the text.

AlexanderLukin committed:
fix: minor cleanups

Minor cleanups based on review notes.

AlexanderLukin committed:
refactor: remove some internal methods

Remove some internal methods and structures from code examples.

AlexanderLukin committed:
refactor: add links to LIP-10

- Add links to the LIP-10 to all appropriate sections.
- Minor text fixes.

AlexanderLukin committed:
refactor: remove byte code optimization parts

- Remove mentions of `ZeroAddressLido` and `ZeroAddressAdmin` error changes from section 2.2.
- Remove the `initialize` method from code examples.

AlexanderLukin committed:
chore: minor text fixes

AlexanderLukin committed:
fix: minor text fixes

AlexanderLukin committed:
feat: new Links section

Add all relevant links.

infloop commented:
Renamed branch to `feat/lip-25` and created another PR https://github.com/lidofinance/lido-improvement-proposals/pull/59
